kong/clustering.lua

```diff
@@ -0,0 +1,409 @@
+    return
+  end
+
+  -- connection established
+  -- ping thread
+  ngx.thread.spawn(function()
+    while true do
+      send_ping(c)
+      ngx_sleep(PING_INTERVAL)
+    end
+  end)
+
+  while true do
+    local data, typ, err = c:recv_frame()
+    if err then
+      ngx.log(ngx.ERR, "error while receiving frame from control plane: ", err)
+      c:close()
+
+      local delay = 9 + math.random()
+      assert(ngx.timer.at(delay, communicate, conf))
+      return
+    end
+
+    if typ == "binary" then
+      local msg = assert(cjson_decode(data))
+
+      if msg.type == "reconfigure" then
+        local config_table = assert(msg.config_table)
+
+        local res, err = update_config(config_table, true)
+        if not res then
+          ngx_log(ngx_ERR, "unable to update running config: ", err)
+        end
+
+        send_ping(c)
+
+      end
+    elseif typ == "pong" then
+      ngx_log(ngx_DEBUG, "received PONG frame from control plane")
+    end
+  end
+end
+
+
+function _M.handle_cp_websocket()
+  -- use mutual TLS authentication
+  local cert = ngx_var.ssl_client_raw_cert
+
+  if not cert then
+    ngx_log(ngx_ERR, "Data Plane failed to present client certificate " ..
+                     "during handshake")
+    return ngx_exit(444)
+  end
+
+  cert = assert(openssl_x509.new(cert, "PEM"))
+  local digest = cert:digest("sha256")
+
+  if CERT_DIGEST and digest ~= CERT_DIGEST then
+    ngx_log(ngx_ERR, "Data Plane presented incorrect client certificate " ..
+                     "during handshake, expected digest: " .. CERT_DIGEST ..
+                     " got: " .. digest)
+    return ngx_exit(444)
+  end
+
+  local node_id = ngx_var.arg_node_id
+  if not node_id then
+    ngx_exit(400)
+  end
+
+  local node_hostname = ngx_var.arg_node_hostname
+  local node_ip = ngx_var.remote_addr
+
+  local wb, err = ws_server:new(WS_OPTS)
+  if not wb then
+    ngx_log(ngx_ERR, "failed to perform server side WebSocket handshake: ", err)
+    return ngx_exit(444)
+  end
+
+  local sem = semaphore.new()
+  local queue = { sem = sem, }
+  clients[wb] = queue
+
+  local res
+  -- unconditionally send config update to new clients to
+  -- ensure they have latest version running
+  res, err = declarative.export_config()
+  if not res then
+    ngx_log(ngx_ERR, "unable to export config from database: ".. err)
+  end
+
+  table.insert(queue, res)
+  queue.sem:post()
+
+  -- connection established
+  -- ping thread
+  ngx.thread.spawn(function()
+    while true do
+      local data, typ, err = wb:recv_frame()
+      if not data then
+        ngx_log(ngx_ERR, "did not receive ping frame from data plane: ", err)
+        -- return and let the main thread handle the error
+        return
+      end
+
+      assert(typ == "ping")
+      local _
+      _, err = wb:send_pong()
+      if err then
+        ngx_log(ngx_ERR, "failed to send PONG back to data plane: ", err)
+        -- return and let the main thread handle the error
```

Comment:
same here, again both bad and good case return nil (as the only return argument).